ml-agents-envs/mlagents/envs/timers.py

```diff
@@ -0,0 +1,172 @@
+# # Unity ML-Agents Toolkit
+from time import perf_counter
+
+from contextlib import contextmanager
+from typing import Any, Callable, Dict, Generator, TypeVar
+
+"""
+Lightweight, hierarchical timers for profiling sections of code.
+
+Example:
+
+@timed
+def foo(t):
+    time.sleep(t)
+
+def main():
+    for i in range(3):
+        foo(i + 1)
+    with hierarchical_timer("context"):
+        foo(1)
+
+    print(get_timer_tree())
+
+This would produce a timer tree like
+    (root)
+        "foo"
+        "context"
+            "foo"
+
+The total time and counts are tracked for each block of code; in this example "foo" and "context.foo" are considered
+distinct blocks, and are tracked separately.
+"""
+
+
+class TimerNode:
+    """
+    Represents the time spent in a block of code.
+    """
+
+    __slots__ = ["children", "total", "count"]
+
+    def __init__(self):
+        self.children: Dict[str, "TimerNode"] = {}
+        self.total: float = 0.0
+        self.count: int = 0
+
+    def get_child(self, name: str) -> "TimerNode":
```

Comment:
Maybe this is a dumb question.. Why do you make the TimeNode into "TimeNode" here?

Reply:
This explains it better than I can: https://mypy.readthedocs.io/en/latest/kinds_of_types.html#class-name-forward-references

The runtime can't parse without the string: ```
______________________________________ ERROR collecting mlagents/trainers/tests/test_reward_signals.py ______________________________________
venv/lib/python3.7/site-packages/py/_path/local.py:701: in pyimport
    __import__(modname)
ml-agents/mlagents/trainers/__init__.py:6: in <module>
    from .trainer import *
ml-agents/mlagents/trainers/trainer.py:8: in <module>
    from mlagents.trainers import TrainerMetrics
ml-agents/mlagents/trainers/__init__.py:8: in <module>
    from .trainer_controller import *
ml-agents/mlagents/trainers/trainer_controller.py:16: in <module>
    from mlagents.envs.subprocess_env_manager import SubprocessEnvManager
ml-agents-envs/mlagents/envs/subprocess_env_manager.py:9: in <module>
    from mlagents.envs.timers import timed, hierarchical_timer
ml-agents-envs/mlagents/envs/timers.py:35: in <module>
    class TimerNode:
ml-agents-envs/mlagents/envs/timers.py:47: in TimerNode
    def get_child(self, name: str) -> TimerNode:
E   NameError: name 'TimerNode' is not defined
```
It looks like they're not needed in `__init__` though, so I'll remove those ones.